drivers/network/ndis/ndis/hardware.c

```diff
@@ -13,6 +13,73 @@
  */
 
 #include "ndissys.h"
+#include <wdmguid.h>
+
+static NTSTATUS
+NdisQueryPciBusInterface(
+    IN PLOGICAL_ADAPTER Adapter)
+{
+    KEVENT Event;
+    NTSTATUS Status;
+    PIRP Irp;
+    IO_STATUS_BLOCK IoStatusBlock;
+    PIO_STACK_LOCATION IrpStack;
+    PBUS_INTERFACE_STANDARD BusInterfacePtr;
+
+    if (Adapter->BusInterfaceQueried) {
+      return (Adapter->BusInterface.GetBusData != NULL) ? STATUS_SUCCESS : STATUS_NOT_SUPPORTED;
+    }
+
+    Adapter->BusInterfaceQueried = TRUE;
+    if (Adapter->NdisMiniportBlock.BusType != NdisInterfacePci) {
+      return STATUS_NOT_SUPPORTED;
+    }
+
+    BusInterfacePtr = ExAllocatePoolWithTag(NonPagedPool, sizeof(BUS_INTERFACE_STANDARD), NDIS_TAG);
+    if (BusInterfacePtr == NULL) {
+      return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    KeInitializeEvent(&Event, NotificationEvent, FALSE);
+
+    Irp = IoBuildSynchronousFsdRequest(IRP_MJ_PNP,
+                       Adapter->NdisMiniportBlock.NextDeviceObject,
+                       NULL,
+                       0,
+                       NULL,
+                       &Event,
+                       &IoStatusBlock);
+    if (Irp == NULL) {
+      ExFreePoolWithTag(BusInterfacePtr, NDIS_TAG);
+      return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    IrpStack = IoGetNextIrpStackLocation(Irp);
+    IrpStack->MajorFunction = IRP_MJ_PNP;
+    IrpStack->MinorFunction = IRP_MN_QUERY_INTERFACE;
+    IrpStack->Parameters.QueryInterface.InterfaceType = &GUID_BUS_INTERFACE_STANDARD;
+    IrpStack->Parameters.QueryInterface.Size = sizeof(BUS_INTERFACE_STANDARD);
+    IrpStack->Parameters.QueryInterface.Version = 1;
+    IrpStack->Parameters.QueryInterface.Interface = (PINTERFACE)BusInterfacePtr;
```

Comment:
I actually wonder whether, instead of allocating the temporary interface pointer, it may be better to just pass here a pointer to the adapter bus interface structure:
```suggestion
    IrpStack->Parameters.QueryInterface.Interface = (PINTERFACE)&Adapter->BusInterface;
```
so that it gets directly initialized in case of success, and in case of failure you would do at most a `RtlZeroMemory(&Adapter->BusInterface, sizeof(Adapter->BusInterface));` on it.